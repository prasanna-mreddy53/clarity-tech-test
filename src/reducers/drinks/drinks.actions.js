import { drinksActionTypes } from "./drinks.types";

export const fetchDrinksRequest = () => ({
  type: drinksActionTypes.FETCH_DRINKS_REQUEST,
});

export const fetchDrinksSuccess = (json) => ({
  type: drinksActionTypes.FETCH_DRINKS_SUCCESS,
  payload: json,
});

export const filterDrinks = (searchTerm) => ({
  type: drinksActionTypes.FILTER_DRINKS,
  payload: searchTerm,
});

export const fetchDrinksFailure = (error) => ({
  type: drinksActionTypes.FETCH_DRINKS_FAILURE,
  payload: error,
});

export const setDrinkCategory = (drink) => ({
  type: drinksActionTypes.SET_DRINK_CATEGORY,
  payload: drink,
});

export const fetchDrinks = (drink) => {
  return async (dispatch) => {
    dispatch(fetchDrinksRequest);
    try {
      let response = await fetch(
        `https://www.thecocktaildb.com/api/json/v1/1/filter.php?c=${drink}`
      );
      let json = await response.json();
      dispatch(fetchDrinksSuccess(json.drinks));
      dispatch(setDrinkCategory(drink));
    } catch (error) {
      dispatch(fetchDrinksFailure(error));
    }
  };
};
