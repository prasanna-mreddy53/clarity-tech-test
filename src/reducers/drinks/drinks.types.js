export const drinksActionTypes = {
  FETCH_DRINKS_REQUEST: "FETCH_DRINKS_REQUEST",
  FETCH_DRINKS_SUCCESS: "FETCH_DRINKS_SUCCESS",
  FETCH_DRINKS_FAILURE: "FETCH_DRINKS_FAILURE",
  FILTER_DRINKS: "FILTER_DRINKS",
  SET_DRINK_CATEGORY: "SET_DRINK_CATEGORY",
};
