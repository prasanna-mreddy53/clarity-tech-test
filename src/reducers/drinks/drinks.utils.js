export const filterDrinks = (drinks, searchTerm) => {
  let updatedDrinks = drinks;
  return updatedDrinks.filter((drink) => {
    return drink.strDrink.toLowerCase().search(searchTerm.toLowerCase()) !== -1;
  });
};
