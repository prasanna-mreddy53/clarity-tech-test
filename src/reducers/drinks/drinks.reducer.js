import { drinksActionTypes } from "./drinks.types";
import { filterDrinks } from "./drinks.utils";

const INITIAL_STATE = {
  loading: false,
  drinks: [],
  filteredDrinks: [],
  error: "",
  drinkCategory: "Ordinary Drink",
};

const drinksReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case drinksActionTypes.FETCH_DRINKS_REQUEST:
      return {
        ...state,
        loading: true,
      };
    case drinksActionTypes.FETCH_DRINKS_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case drinksActionTypes.FETCH_DRINKS_SUCCESS:
      return {
        ...state,
        drinks: action.payload,
        filteredDrinks: action.payload,
      };
    case drinksActionTypes.FILTER_DRINKS:
      return {
        ...state,
        filteredDrinks: filterDrinks(state.drinks, action.payload),
      };
    case drinksActionTypes.SET_DRINK_CATEGORY:
      return {
        ...state,
        drinkCategory: action.payload,
      };
    default:
      return state;
  }
};

export default drinksReducer;
