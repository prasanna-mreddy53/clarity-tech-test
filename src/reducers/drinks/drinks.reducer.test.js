import reducer from "./drinks.reducer";
import { drinksActionTypes } from "./drinks.types";

describe("Drinks reducer", () => {
  it("Drinks should return the initial state", () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it("should return the initial state with default values", () => {
    expect(reducer(undefined, {})).toEqual({
      loading: false,
      drinks: [],
      filteredDrinks: [],
      error: "",
      drinkCategory: "Ordinary Drink",
    });
  });

  it("should match FETCH_DRINKS_REQUEST action", () => {
    expect(
      reducer([], {
        type: drinksActionTypes.FETCH_DRINKS_REQUEST,
      })
    ).toEqual({
      loading: true,
    });
  });

  it("should match FETCH_DRINKS_SUCCESS action", () => {
    expect(
      reducer(
        { drinks: [], loading: false, error: "" },
        {
          type: drinksActionTypes.FETCH_DRINKS_SUCCESS,
          drinks: { drinks: [{ strDrink: "Margarita" }] },
        }
      )
    ).toEqual({
      drinks: [],
      loading: false,
      error: "",
      drinks: undefined,
    });
  });

  it("Drinks FETCH_DRINKS_SUCCESS action", () => {
    expect(
      reducer(
        {
          drinks: { drinks: [{ strDrink: "Margarita" }] },
        },
        drinksActionTypes.FETCH_DRINKS_SUCCESS
      )
    ).toMatchSnapshot();
  });

  it("should match FETCH_DRINKS_FAILURE action", () => {
    expect(
      reducer([{ drinks: [], loading: false, error: "" }], {
        type: drinksActionTypes.FETCH_DRINKS_FAILURE,
        error: "could not load drinks",
      })
    ).toMatchSnapshot();
  });
});
