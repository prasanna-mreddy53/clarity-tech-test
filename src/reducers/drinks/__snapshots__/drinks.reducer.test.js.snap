// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drinks reducer Drinks FETCH_DRINKS_SUCCESS action 1`] = `
Object {
  "drinks": Object {
    "drinks": Array [
      Object {
        "strDrink": "Margarita",
      },
    ],
  },
}
`;

exports[`Drinks reducer Drinks should return the initial state 1`] = `
Object {
  "drinkCategory": "Ordinary Drink",
  "drinks": Array [],
  "error": "",
  "filteredDrinks": Array [],
  "loading": false,
}
`;

exports[`Drinks reducer should match FETCH_DRINKS_FAILURE action 1`] = `
Object {
  "0": Object {
    "drinks": Array [],
    "error": "",
    "loading": false,
  },
  "error": undefined,
}
`;
